dnsproxy: introduce cache to reference count regexes

odinuge opened the pull request:
This introduce a reference counted cache/index for regexes, and is ready for supporting other caches as well. This means that policies with the same set of rules can reuse the underlying matcher saving, since long regex tends to occupy a lot of memory.

This also makes the LRU cache for compiled regexes more or less redundant, mitigating the drawbacks of that cache. Having a large LRU cache is useful when a lot of similar policies are in use at the same time, but in case a node have a small set of policies inserted at the same time, but the churn makes new and distinct policies come and go, the LRU cache will use a lot more memory than the benefit it gives.

benchmark TL;DR: Heap usage compared to a LRU size of 128 is down by ~99%, and ~92% with a LRU size of 1024, with the given test data*.

* defining heap usage as (HeapInuse after test) - (HeapInuse before setup)
```
$ go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x ./pkg/fqdn/dnsproxy

// With this change;

go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x -memprofile memprofile.out ./pkg/fqdn/dnsproxy goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=128)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 20 MiB    NumGC = 6
Before Test (N=1,EPs=20,cache=128)
Alloc = 37 MiB  HeapInuse = 49 MiB      Sys = 1031 MiB  NumGC = 18
After Test (N=1,EPs=20,cache=128)
Alloc = 52 MiB  HeapInuse = 62 MiB      Sys = 1031 MiB  NumGC = 53
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        4289436000 ns/op        2559832616 B/op 34303565 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      12.574s

// Baseline with LRU size of 1024

goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=1024)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 20 MiB    NumGC = 6
Before Test (N=1,EPs=20,cache=1024)
Alloc = 37 MiB  HeapInuse = 49 MiB      Sys = 969 MiB   NumGC = 18
After Test (N=1,EPs=20,cache=1024)
Alloc = 312 MiB HeapInuse = 357 MiB     Sys = 969 MiB   NumGC = 46
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        5990697400 ns/op        3470756528 B/op 36068464 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      11.813s

// Baseline with LRU size of 128
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=128)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 19 MiB    NumGC = 5
Before Test (N=1,EPs=20,cache=128)
Alloc = 38 MiB  HeapInuse = 49 MiB      Sys = 1047 MiB  NumGC = 17
After Test (N=1,EPs=20,cache=128)
Alloc = 1763 MiB        HeapInuse = 1907 MiB    Sys = 2918 MiB  NumGC = 44
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        81058653300 ns/op       10328100808 B/op        46896658 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      93.574s
```

```release-note
Introduce smarter internal cache to reduce memory consumption for FQDN / DNS policy usage, especially in environment with heavy FQDN / DNS policy usage
```

odinuge-work committed:
dnsproxy: introduce matcher cache to reference count regexes

This introduce a reference counted cache/index for regexes, and is ready for
supporting other caches as well. This means that policies with the same set
of rules can reuse the underlying matcher saving, since long regex tends
to occupy a lot of memory.

This also makes the LRU cache for compiled regexes more or less
redundant, mitigating the drawbacks of that cache. Having a large LRU
cache is useful when a lot of similar policies are in use at the same
time, but in case a node have a small set of policies inserted at the
same time, but the churn makes new and distinct policies come and go,
the LRU cache will use a lot more memory than the benefit it gives.

benchmark TL;DR: Heap usage compared to a LRU size of 128 is down by
~99%, and ~92% with a LRU size of 1024, with the given test data*.

* defining heap usage as (HeapInuse after test) - (HeapInuse before setup)

$ benchstat cache-1024-entries.txt pr-21895-without-cache.txt
name                             old time/op          new time/op          delta
_perEPAllow_setPortRulesForID-7           83.7s ±17%           77.4s ± 8%     ~     (p=0.310 n=5+5)

name                             old B(HeapInUse)/op  new B(HeapInUse)/op  delta
_perEPAllow_setPortRulesForID-7           1.79G ± 0%           0.10G ± 1%  -94.55%  (p=0.008 n=5+5)

name                             old alloc/op         new alloc/op         delta
_perEPAllow_setPortRulesForID-7          78.6GB ± 0%          78.6GB ± 0%   -0.00%  (p=0.008 n=5+5)

name                             old allocs/op        new allocs/op        delta
_perEPAllow_setPortRulesForID-7            167M ± 0%            167M ± 0%   -0.01%  (p=0.008 n=5+5)

$ go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x ./pkg/fqdn/dnsproxy

// With this change;

go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x -memprofile memprofile.out ./pkg/fqdn/dnsproxy
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=128)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 20 MiB    NumGC = 6
Before Test (N=1,EPs=20,cache=128)
Alloc = 37 MiB  HeapInuse = 49 MiB      Sys = 1031 MiB  NumGC = 18
After Test (N=1,EPs=20,cache=128)
Alloc = 52 MiB  HeapInuse = 62 MiB      Sys = 1031 MiB  NumGC = 53
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        4289436000 ns/op        2559832616 B/op 34303565 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      12.574s

// Baseline with LRU size of 1024

goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=1024)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 20 MiB    NumGC = 6
Before Test (N=1,EPs=20,cache=1024)
Alloc = 37 MiB  HeapInuse = 49 MiB      Sys = 969 MiB   NumGC = 18
After Test (N=1,EPs=20,cache=1024)
Alloc = 312 MiB HeapInuse = 357 MiB     Sys = 969 MiB   NumGC = 46
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        5990697400 ns/op        3470756528 B/op 36068464 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      11.813s

// Baseline with LRU size of 128
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=128)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 19 MiB    NumGC = 5
Before Test (N=1,EPs=20,cache=128)
Alloc = 38 MiB  HeapInuse = 49 MiB      Sys = 1047 MiB  NumGC = 17
After Test (N=1,EPs=20,cache=128)
Alloc = 1763 MiB        HeapInuse = 1907 MiB    Sys = 2918 MiB  NumGC = 44
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        81058653300 ns/op       10328100808 B/op        46896658 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      93.574s

Signed-off-by: Odin Ugedal <odin@uged.al>
Signed-off-by: Odin Ugedal <ougedal@palantir.com>

odinuge commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -157,22 +161,77 @@ type DNSProxy struct {
+
+// CachedSelectorEntry maps port numbers to selectors to rules, mirroring
+// policy.L7DataMap but the DNS rules are compiled into a regex
+type CachedSelectorEntry map[policy.CachedSelector]fqdnMatcher
```
We can flattern this right now to be `*regexp.Regexp` instead of `fqdnMatcher`. If/when we want the map/list of matchNames separately, we can introduce the `fqdnMatcher`

christarazi replied:
Sure, let's flatten it now.

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -157,22 +161,77 @@ type DNSProxy struct {
+		return true
+	}
+	return false
+}
```
```suggestion
	return r.patternMatcher != nil && r.patternMatcher.MatchString(fqdn)
}
```

odinuge replied:
Thanks! Ended up refactoring this out since I removed that `fqdnMatcher` struct

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -274,61 +363,154 @@ func (p *DNSProxy) RemoveRestoredRules(endpointID uint16) {
+// lookupOrCompilePatternMatcher will check if the pattern is already compiled and present in another policy, and
+// will reuse it in order to reduce memory consumption. The usage is reference counted, so all calls where
+// lookupOrCompilePatternMatcher returns no error, a subsequent call to release it via releasePatternMatcher has to
+// be tone when its no longer being used by the policy.
```
```suggestion
// be done when it's no longer being used by the policy.
```

odinuge replied:
Good catch, thanks!

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -274,61 +363,154 @@ func (p *DNSProxy) RemoveRestoredRules(endpointID uint16) {
+	if indexEntry, ok := c.patternMatcherCache[patternMatcher.String()]; ok {
+		switch indexEntry.referenceCount {
+		case 1:
+			delete(c.patternMatcherCache, patternMatcher.String())
```
Let's call `String()` only once so we don't waste cycles doing duplicate work

odinuge replied:
Yeah, sounds good!

christarazi commented on pkg/fqdn/restore/restore.go:
```diff
@@ -21,20 +20,24 @@ type IPRules []IPRule
+	// v1.12 and older versions depend on having this value provided as a string, but for future
+	// improvements we gracefully handle reading it as nil as well. We always ensure that we currently
+	// never write nil in order to break rolling back to v1.12 or earlier.
+	Re  *string
```
Have you tested downgrading from this version and validated that the rules are restored properly and traffic is not disrupted?

odinuge replied:
Ended up "reverting" this change and keeping "99%" of the old logic here to ensure we behave the same way.

The actual disk format is exactly the same, so are the patterns/regexes.

I guess we can write some more tests, tho. I think it should be ok covered by `TestRestoredEndpoint` and `TestFullPathDependence`.

christarazi reviewed (changes requested):
Looks great! A few comments

christarazi commented on pkg/fqdn/restore/restore.go:
```diff
@@ -25,17 +24,24 @@ type IPRule struct {
 
-// RuleRegex is a wrapper for *regexp.Regexp so that we can define marshalers for it.
+// RuleRegex is a wrapper for a pointer to a string so that we can define marshalers for it.
 type RuleRegex struct {
-	*regexp.Regexp
+	Pattern *string
 }
```
I feel like this change along with the changes to the marshaling function can be in their own separate commit. Even though it helps with the overall optimization of introducing a cache, it can still be standalone because it sort of an optimization itself + at the same time being a preparatory commit for the main cache optimization.

odinuge replied:
Yee, agree! Will split them!

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -249,7 +284,27 @@ func (p *DNSProxy) RestoreRules(ep *endpoint.Endpoint) {
+				var err error
+				regex, err = p.cache.lookupOrCompileRegex(*ipRule.Re.Pattern)
+				if err != nil {
+					log.Warnf("Error compiling regex for endpointID %d, gracefully ignoring", ep.ID)
```
I think `INFO` level is more appropriate for this log msg. I think we'll need a more user-friendly msg as well.

```suggestion
					log.WithFields(logrus.Fields{
					  logfields.EndpointID: ep.ID,
					  logfields.Rule: *ipRule.Re.Pattern,
					}).Info("Disregarding restored DNS rule due to failure in compiling regex. Traffic to the FQDN may be disrupted.")
```

odinuge-work committed:
dnsproxy: stop compiling regexes in json marshalling

There is no need to unmarshal regex patterns directly into compiled
regexes. Instead, marshal the pattern into a string that we later
compile in the dnsproxy to avoid duplicate work. This ensures that we
only compile two equal patterns once, ensuring that we save both memory
and cpu time during startup.

Function wise this is a no-op, as long as the regexes are valid. If they
are invalid, we now unmarshal successfully into a string, and
gracefully ignores the provided regex in case we are unable to compile
it. Previously this was handled in the json unmarshaller.

Signed-off-by: Odin Ugedal <ougedal@palantir.com>

odinuge commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -210,8 +214,11 @@ func (p *DNSProxy) checkRestored(endpointID uint64, destPort uint16, destIP stri
-			return true
+		ipRule := ipRules[i]
+		if _, exists := ipRule.IPs[destIP]; exists || ipRule.IPs == nil {
+			if ipRule.regex != nil && ipRule.regex.MatchString(name) {
```
Should I skip this `nil` check so we can still keep it "undefined" until we decide on the actual semantics of a `nil` value? Tho. I think treating `nil` as no match is probably the most meaningful one.

christarazi replied:
Is it possible for it to be `nil` or is this just coming from overly defensive coding? In the end, I think treating the `nil` as no match makes sense regardless.

odinuge replied:
Hmm, yeah, mostly defensive since we discussed this in a previous PR in regards to the list of `matchNames`.

odinuge replied:
> Is it possible for it to be nil

as far as i understand, the only reason for nil is if someone manually change the json file on disk to either nil or `""` (I think), so it should never happen

christarazi replied:
Might as well keep it. We can treat it as no-match. In case someone accidentally mucks with the json file or somehow the data is corrupted, we don't want to inadvertently allow traffic, even though the chances of this happening are extremely small.

odinuge commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -249,7 +284,30 @@ func (p *DNSProxy) RestoreRules(ep *endpoint.Endpoint) {
+	for port, dnsRule := range ep.DNSRules {
+		ipRules := make([]restoredIPRule, 0, len(dnsRule))
+		for _, ipRule := range dnsRule {
+			if ipRule.Re.Pattern == nil {
```
This is also assuming we define a `null` value as not allowed

christarazi commented on pkg/fqdn/dnsproxy/proxy_test.go:
```diff
@@ -17,6 +17,8 @@ import (
 	"time"
 
+	"golang.org/x/exp/maps"
+
```
Nit: no need for extra newline

odinuge replied:
Ahh, goimports moved there... I'll move it down and reformat! Thanks

christarazi commented on pkg/fqdn/dnsproxy/proxy_test.go:
```diff
@@ -1000,6 +1002,35 @@ func (s *DNSProxyTestSuite) TestRestoredEndpoint(c *C) {
 
+	invalidRePattern := "invalid-re-pattern((*"
+	validRePattern := "^this[.]domain[.]com[.]$"
+	k := maps.Keys(restored)[0]
```
`maps.Keys()` returns the keys in an indeterminate order. Does the test make any assmptions about the order of the keys returned? 

```
// Keys returns the keys of the map m.
// The keys will be in an indeterminate order.
```

https://github.com/odinuge/cilium/blob/e28cce148c676371a234a51f303c8c7933f81960/vendor/golang.org/x/exp/maps/maps.go#L10

odinuge replied:
The order doesn't matter since this map only has one element.

I _think_ the key is actually random since its the port of the dns-server, and it listens to `:0` that I think means "any free" port or so. Guess we could parse `s.proxy.TCPServer.Listener.Addr()` as well, but I thought this was the simplest. Maybe I should add a comment about it

ref. https://github.com/cilium/cilium/blob/e28cce148c676371a234a51f303c8c7933f81960/pkg/fqdn/dnsproxy/proxy_test.go#L120

christarazi commented on pkg/fqdn/dnsproxy/proxy_test.go:
```diff
@@ -437,6 +449,32 @@ func (s *DNSProxyTestSuite) TestRespondMixedCaseInRequestResponse(c *C) {
+	}
+	query := name
+
+	err := s.proxy.UpdateAllowed(epID1, dstPort, l7map)
```
Lint check failed here

christarazi commented on pkg/fqdn/dnsproxy/proxy_test.go:
```diff
@@ -437,6 +449,32 @@ func (s *DNSProxyTestSuite) TestRespondMixedCaseInRequestResponse(c *C) {
+			},
+		},
+	}
+	err = s.proxy.UpdateAllowed(epID1, dstPort, l7map)
```
Lint check failed here as well

odinuge replied:
Ooof, will fix, thanks!
